Crash on Newest Version of Fabric

Crash on Fabric 0.12.1
Minecraft Version: 1.17.1

Mods: 
Cloth Config 5.0.38
Arcanus 1.16
Trinkets 3.0.4
Fabric API 0.40.8

Arcanus 1.16 only works currently on Fabric 0.11.7

[14:06:34] [main/INFO]: Compatibility level set to JAVA_16
Failed to start Minecraft:
java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at org.multimc.onesix.OneSixLauncher.launchWithMainClass(OneSixLauncher.java:196)
	at org.multimc.onesix.OneSixLauncher.launch(OneSixLauncher.java:231)
	at org.multimc.EntryPoint.listen(EntryPoint.java:143)
	at org.multimc.EntryPoint.main(EntryPoint.java:34)
Caused by: java.lang.RuntimeException: Minecraft has crashed
	at net.fabricmc.loader.impl.game.minecraft.MinecraftGameProvider.launch(MinecraftGameProvider.java:571)
	at net.fabricmc.loader.impl.launch.knot.Knot.launch(Knot.java:74)
	at net.fabricmc.loader.impl.launch.knot.KnotClient.main(KnotClient.java:23)
	... 8 more
Caused by: java.lang.ExceptionInInitializerError
	at net.minecraft.class_3523.<clinit>(class_3523.java:13)
	at net.minecraft.class_5471.<clinit>(class_5471.java:11)
	at net.minecraft.class_5458.method_30573(class_5458.java:44)
	at net.minecraft.class_5458.method_30566(class_5458.java:75)
	at java.base/java.util.LinkedHashMap.forEach(LinkedHashMap.java:723)
	at net.minecraft.class_5458.<clinit>(class_5458.java:74)
	at net.minecraft.class_2378.<clinit>(class_2378.java:266)
	at net.minecraft.class_2966.method_12851(class_2966.java:44)
	at net.minecraft.client.main.Main.main(Main.java:139)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at net.fabricmc.loader.impl.game.minecraft.MinecraftGameProvider.launch(MinecraftGameProvider.java:568)
	... 10 more
Caused by: java.lang.RuntimeException: Mixin transformation of net.minecraft.class_1309 failed
	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.getPostMixinClassByteArray(KnotClassDelegate.java:224)
	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.tryLoadClass(KnotClassDelegate.java:133)
	at net.fabricmc.loader.impl.launch.knot.KnotClassLoader.loadClass(KnotClassLoader.java:155)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:519)
	at java.base/java.lang.ClassLoader.defineClass1(Native Method)
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1010)
	at java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:150)
	at net.fabricmc.loader.impl.launch.knot.KnotClassLoader.defineClassFwd(KnotClassLoader.java:218)
	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.tryLoadClass(KnotClassDelegate.java:149)
	at net.fabricmc.loader.impl.launch.knot.KnotClassLoader.loadClass(KnotClassLoader.java:155)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:519)
	at net.minecraft.class_2246.<clinit>(class_2246.java:103)
	... 24 more
Caused by: org.spongepowered.asm.mixin.transformer.throwables.MixinTransformerError: An unexpected critical error was encountered
	at org.spongepowered.asm.mixin.transformer.MixinProcessor.applyMixins(MixinProcessor.java:392)
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.transformClass(MixinTransformer.java:234)
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.transformClassBytes(MixinTransformer.java:202)
	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.getPostMixinClassByteArray(KnotClassDelegate.java:222)
	... 35 more
Caused by: org.spongepowered.asm.mixin.injection.throwables.InjectionError: Critical injection failure: Variable modifier method damage(FLnet/minecraft/class_1282;)F in arcanus.mixins.json:LivingEntityMixin from mod arcanus failed injection check, (0/1) succeeded. Scanned 1 target(s). Using refmap arcanus-refmap.json Messages: { @At("LOAD" implicit float) has invalid IMPLICIT discriminator for opcode 577 in net/minecraft/class_1309::method_5643(Lnet/minecraft/class_1282;F)Z: Found 3 candidate variables but exactly 1 is required. @At("LOAD" implicit float) has invalid IMPLICIT discriminator for opcode 619 in net/minecraft/class_1309::method_5643(Lnet/minecraft/class_1282;F)Z: Found 3 candidate variables but exactly 1 is required. @At("LOAD" implicit float) has invalid IMPLICIT discriminator for opcode 620 in net/minecraft/class_1309::method_5643(Lnet/minecraft/class_1282;F)Z: Found 3 candidate variables but exactly 1 is required. @At("LOAD" implicit float) has invalid IMPLICIT discriminator for opcode 625 in net/minecraft/class_1309::method_5643(Lnet/minecraft/class_1282;F)Z: Found 3 candidate variables but exactly 1 is required. @At("LOAD" implicit float) has invalid IMPLICIT discriminator for opcode 629 in net/minecraft/class_1309::method_5643(Lnet/minecraft/class_1282;F)Z: Found 3 candidate variables but exactly 1 is required. @At("LOAD" implicit float) has invalid IMPLICIT discriminator for opcode 638 in net/minecraft/class_1309::method_5643(Lnet/minecraft/class_1282;F)Z: Found 3 candidate variables but exactly 1 is required. @At("LOAD" implicit float) has invalid IMPLICIT discriminator for opcode 656 in net/minecraft/class_1309::method_5643(Lnet/minecraft/class_1282;F)Z: Found 3 candidate variables but exactly 1 is required. @At("LOAD" implicit float) has invalid IMPLICIT discriminator for opcode 657 in net/minecraft/class_1309::method_5643(Lnet/minecraft/class_1282;F)Z: Found 3 candidate variables but exactly 1 is required.}
	at org.spongepowered.asm.mixin.injection.struct.InjectionInfo.postInject(InjectionInfo.java:468)
	at org.spongepowered.asm.mixin.transformer.MixinTargetContext.applyInjections(MixinTargetContext.java:1381)
	at org.spongepowered.asm.mixin.transformer.MixinApplicatorStandard.applyInjections(MixinApplicatorStandard.java:1052)
	at org.spongepowered.asm.mixin.transformer.MixinApplicatorStandard.applyMixin(MixinApplicatorStandard.java:400)
	at org.spongepowered.asm.mixin.transformer.MixinApplicatorStandard.apply(MixinApplicatorStandard.java:325)
	at org.spongepowered.asm.mixin.transformer.TargetClassContext.apply(TargetClassContext.java:421)
	at org.spongepowered.asm.mixin.transformer.TargetClassContext.applyMixins(TargetClassContext.java:403)
	at org.spongepowered.asm.mixin.transformer.MixinProcessor.applyMixins(MixinProcessor.java:363)
	... 38 more
Exiting with -1
Process exited with code -1.

--- a/gradle.properties
+++ b/gradle.properties
@@ -3,20 +3,20 @@ org.gradle.jvmargs=-Xmx1G
 
 # Modrinth Shenanigans
 modrinth_id=zgaWtmrV
-modrinth_game_versions=1.17,1.17.1
+modrinth_game_versions=1.17.1
 
 # Fabric Properties
 minecraft_version=1.17.1
-yarn_mappings=1.17.1+build.30
-loader_version=0.11.6
+yarn_mappings=1.17.1+build.61
+loader_version=0.12.1
 
 # Mod Properties
 maven_group=dev.cammiescorner
 archives_base_name=arcanus
 
 # Dependencies
-fabric_version=0.40.1+1.17
+fabric_version=0.40.8+1.17
 cloth_config_version=5.0.38
 trinkets_version=3.0.4
-modmenu_version=2.0.10
+modmenu_version=2.0.14
 asa_version=1.2

--- a/src/main/java/dev/cammiescorner/arcanus/core/mixin/ItemStackMixin.java
+++ b/src/main/java/dev/cammiescorner/arcanus/core/mixin/ItemStackMixin.java
@@ -1,23 +1,16 @@
 package dev.cammiescorner.arcanus.core.mixin;
 
-import com.google.common.collect.Multimap;
-import net.minecraft.client.item.TooltipContext;
-import net.minecraft.entity.EquipmentSlot;
 import net.minecraft.entity.attribute.EntityAttribute;
 import net.minecraft.entity.attribute.EntityAttributeModifier;
-import net.minecraft.entity.player.PlayerEntity;
 import net.minecraft.item.ItemStack;
-import net.minecraft.text.MutableText;
-import net.minecraft.text.Text;
 import net.minecraft.util.Formatting;
 import org.spongepowered.asm.mixin.Mixin;
 import org.spongepowered.asm.mixin.Unique;
 import org.spongepowered.asm.mixin.injection.At;
-import org.spongepowered.asm.mixin.injection.Inject;
-import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
-import org.spongepowered.asm.mixin.injection.callback.LocalCapture;
+import org.spongepowered.asm.mixin.injection.ModifyArg;
+import org.spongepowered.asm.mixin.injection.ModifyVariable;
+import org.spongepowered.asm.mixin.injection.Slice;
 
-import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
 
@@ -26,16 +19,21 @@ import static dev.cammiescorner.arcanus.Arcanus.EntityAttributes.*;
 @Mixin(ItemStack.class)
 public class ItemStackMixin {
 	@Unique private final List<EntityAttribute> inverseAttributes = List.of(MANA_COST, MANA_REGEN, BURNOUT_REGEN, MANA_LOCK);
+	@Unique private boolean affectCurrentAttribute;
 
-	@Inject(method = "getTooltip", at = @At(value = "INVOKE_ASSIGN", target = "Ljava/util/List;add(Ljava/lang/Object;)Z", ordinal = 8), locals = LocalCapture.CAPTURE_FAILSOFT)
-	public void changePositiveFormatting(PlayerEntity player, TooltipContext context, CallbackInfoReturnable<List<Text>> info, List<Text> list, int i, EquipmentSlot[] slots, int j, int k, EquipmentSlot slot, Multimap<EntityAttribute, EntityAttributeModifier> map, Iterator<EntityAttributeModifier> iterator, Map.Entry<EntityAttribute, EntityAttributeModifier> entry, EntityAttributeModifier modifier, double d, double g) {
-		if(inverseAttributes.contains(entry.getKey()))
-			((MutableText) list.get(list.size() - 1)).formatted(Formatting.RED);
+	@ModifyVariable(method = "getTooltip", slice = @Slice(from = @At(value = "INVOKE", target = "Ljava/util/Iterator;next()Ljava/lang/Object;")), at = @At(value = "STORE"), ordinal = 0)
+	private Map.Entry<EntityAttribute, EntityAttributeModifier> captureEntry(Map.Entry<EntityAttribute, EntityAttributeModifier> entry) {
+		affectCurrentAttribute = inverseAttributes.contains(entry.getKey());
+		return entry;
 	}
 
-	@Inject(method = "getTooltip", at = @At(value = "INVOKE_ASSIGN", target = "Ljava/util/List;add(Ljava/lang/Object;)Z", ordinal = 9), locals = LocalCapture.CAPTURE_FAILSOFT)
-	public void changeNegativeFormatting(PlayerEntity player, TooltipContext context, CallbackInfoReturnable<List<Text>> info, List<Text> list, int i, EquipmentSlot[] slots, int j, int k, EquipmentSlot slot, Multimap<EntityAttribute, EntityAttributeModifier> map, Iterator<EntityAttributeModifier> iterator, Map.Entry<EntityAttribute, EntityAttributeModifier> entry, EntityAttributeModifier modifier, double d, double g) {
-		if(inverseAttributes.contains(entry.getKey()))
-			((MutableText) list.get(list.size() - 1)).formatted(Formatting.BLUE);
+	@ModifyArg(method = "getTooltip", slice = @Slice(from = @At(value = "CONSTANT", args = "doubleValue=0.0", ordinal = 0)), at = @At(value = "INVOKE", target = "Lnet/minecraft/text/TranslatableText;formatted(Lnet/minecraft/util/Formatting;)Lnet/minecraft/text/MutableText;", ordinal = 0))
+	private Formatting changePositiveFormatting(Formatting value) {
+		return affectCurrentAttribute ? Formatting.RED : value;
+	}
+
+	@ModifyArg(method = "getTooltip", slice = @Slice(from = @At(value = "CONSTANT", args = "doubleValue=0.0", ordinal = 1)), at = @At(value = "INVOKE", target = "Lnet/minecraft/text/TranslatableText;formatted(Lnet/minecraft/util/Formatting;)Lnet/minecraft/text/MutableText;", ordinal = 0))
+	private Formatting changeNegativeFormatting(Formatting value) {
+		return affectCurrentAttribute ? Formatting.BLUE : value;
 	}
 }

--- a/src/main/java/dev/cammiescorner/arcanus/core/mixin/LecternBlockMixin.java
+++ b/src/main/java/dev/cammiescorner/arcanus/core/mixin/LecternBlockMixin.java
@@ -1,27 +1,19 @@
 package dev.cammiescorner.arcanus.core.mixin;
 
-import dev.cammiescorner.arcanus.core.util.MagicUser;
+import dev.cammiescorner.arcanus.core.util.ArcanusHelper;
 import net.minecraft.block.LecternBlock;
-import net.minecraft.block.entity.BlockEntity;
-import net.minecraft.block.entity.LecternBlockEntity;
 import net.minecraft.entity.player.PlayerEntity;
-import net.minecraft.item.ItemStack;
-import net.minecraft.util.Identifier;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
 import org.spongepowered.asm.mixin.Mixin;
 import org.spongepowered.asm.mixin.injection.At;
 import org.spongepowered.asm.mixin.injection.Inject;
 import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
-import org.spongepowered.asm.mixin.injection.callback.LocalCapture;
 
 @Mixin(LecternBlock.class)
 public class LecternBlockMixin{
-	@Inject(method = "openScreen", at = @At(value = "INVOKE", target = "Lnet/minecraft/entity/player/PlayerEntity;openHandledScreen(Lnet/minecraft/screen/NamedScreenHandlerFactory;)Ljava/util/OptionalInt;"), locals = LocalCapture.CAPTURE_FAILSOFT)
-	public void openScreen(World world, BlockPos pos, PlayerEntity player, CallbackInfo info, BlockEntity blockEntity) {
-		ItemStack stack = ((LecternBlockEntity) blockEntity).getBook();
-
-		if(!world.isClient() && stack.getOrCreateNbt().contains("spell"))
-			((MagicUser) player).setKnownSpell(new Identifier(stack.getOrCreateNbt().getString("spell")));
+	@Inject(method = "openScreen", at = @At(value = "RETURN")) // have to do it this way, else the max JVM stack size is exceeded
+	public void openScreen(World world, BlockPos pos, PlayerEntity player, CallbackInfo info) {
+		ArcanusHelper.interactLecternBlock(world, pos, player);
 	}
 }

--- a/src/main/java/dev/cammiescorner/arcanus/core/mixin/LivingEntityMixin.java
+++ b/src/main/java/dev/cammiescorner/arcanus/core/mixin/LivingEntityMixin.java
@@ -51,7 +51,7 @@ public abstract class LivingEntityMixin extends Entity implements CanBeDiscombob
 			setDiscombobulated(false);
 	}
 
-	@ModifyVariable(method = "damage", slice = @Slice(from = @At(value = "INVOKE", target = "Lnet/minecraft/entity/LivingEntity;isDead()Z", ordinal = 1)), at = @At(value = "LOAD", ordinal = 0))
+	@ModifyVariable(method = "damage", slice = @Slice(from = @At(value = "INVOKE", target = "Lnet/minecraft/entity/LivingEntity;isDead()Z", ordinal = 1)), at = @At(value = "LOAD", ordinal = 0), ordinal = 0, argsOnly = true)
 	public float damage(float amount, DamageSource source) {
 		return ArcanusHelper.trinketOnAttack(source, amount, (LivingEntity) (Object) this);
 	}

--- a/src/main/java/dev/cammiescorner/arcanus/core/util/ArcanusHelper.java
+++ b/src/main/java/dev/cammiescorner/arcanus/core/util/ArcanusHelper.java
@@ -4,6 +4,7 @@ import dev.cammiescorner.arcanus.Arcanus;
 import dev.cammiescorner.arcanus.common.items.trinkets.ArcanusTrinketItem;
 import dev.cammiescorner.arcanus.core.registry.ModItems;
 import dev.emi.trinkets.api.TrinketsApi;
+import net.minecraft.block.entity.LecternBlockEntity;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.LivingEntity;
 import net.minecraft.entity.attribute.EntityAttributeInstance;
@@ -12,12 +13,17 @@ import net.minecraft.entity.player.PlayerEntity;
 import net.minecraft.entity.projectile.ProjectileUtil;
 import net.minecraft.item.ItemStack;
 import net.minecraft.nbt.NbtCompound;
+import net.minecraft.nbt.NbtElement;
 import net.minecraft.particle.ParticleEffect;
+import net.minecraft.screen.LecternScreenHandler;
 import net.minecraft.server.world.ServerWorld;
+import net.minecraft.util.Identifier;
 import net.minecraft.util.hit.HitResult;
+import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.Vec3d;
 import net.minecraft.world.RaycastContext;
 import net.minecraft.world.World;
+import org.jetbrains.annotations.ApiStatus;
 import org.jetbrains.annotations.Nullable;
 
 import java.util.List;
@@ -126,5 +132,14 @@ public class ArcanusHelper {
 		tag = stack.getOrCreateSubNbt(Arcanus.MOD_ID);
 		tag.putInt("Exp", 6400);
 		list.add(stack);
+	}
+
+	@ApiStatus.Internal
+	public static void interactLecternBlock(World world, BlockPos pos, PlayerEntity player) {
+		if(!world.isClient() && world.getBlockEntity(pos) instanceof LecternBlockEntity lectern && player.currentScreenHandler instanceof LecternScreenHandler) {
+			NbtCompound nbt = lectern.getBook().getNbt();
+			if (nbt != null && nbt.contains("spell", NbtElement.STRING_TYPE))
+				((MagicUser) player).setKnownSpell(new Identifier(nbt.getString("spell")));
+		}
 	}
 }

--- a/src/main/java/dev/cammiescorner/arcanus/core/util/ArcanusLootFunction.java
+++ b/src/main/java/dev/cammiescorner/arcanus/core/util/ArcanusLootFunction.java
@@ -27,7 +27,7 @@ public class ArcanusLootFunction extends ConditionalLootFunction {
 
 	public static class Builder extends ConditionalLootFunction.Builder<ArcanusLootFunction.Builder> {
 		@Override
-		protected Builder getThisBuilder() {
+		protected ArcanusLootFunction.Builder getThisBuilder() {
 			return this;
 		}
 

--- a/src/main/resources/fabric.mod.json
+++ b/src/main/resources/fabric.mod.json
@@ -30,13 +30,10 @@
 	],
 	"accessWidener": "META-INF/arcanus.accesswidener",
 	"depends": {
-		"fabricloader": ">=0.7.4",
+		"fabricloader": ">=0.12.0-",
 		"fabric": "*",
 		"minecraft": "1.17.x",
 		"cloth-config2": ">=5.0.38",
 		"trinkets": ">=3.0.4"
-	},
-	"suggests": {
-		"flamingo": "*"
 	}
 }

